CreateSomaticPanelOfNormals (i) counts same sample twice, (ii) misses a site and (iii) does not take a list of vcfs

I bring the first point up given the safeguards our previous workflow had in place. In GATK3, CombineVariants checks that all the sample names are unique. You have to add `--genotypemergeoption UNIQUIFY` to allow use of the same normal sample.

In GATK4, CreateSomaticPanelOfNormals allows input of the same sample twice and thereby counts the same evidence in the PoN it generates. Granted, I am doing something intentionally stupid here but I can imagine that this could happen accidentally for situations where we are dealing with hundreds of normal samples. 

Interestingly, given the same sample twice, for my test case, it generates a PoN that has one less call than the inputs.

Here is the command that runs successfully. Note that 2 is a direct copy of 1 made using unix `cp`.
```
	gatk-launch CreateSomaticPanelOfNormals \
		-vcfs 1_normalforpon.vcf.gz \
		-vcfs 2_normalforpon.vcf.gz \
		-O 3_discard_practice_pon.vcf.gz
```

Then I check the number of sites and I see a discrepency.
```
WMCF9-CB5:working shlee$ gzcat 3_discard_practice_pon.vcf.gz | grep -v '#' | wc -l
     138
WMCF9-CB5:working shlee$ gzcat 1_normalforpon.vcf.gz | grep -v '#' | wc -l
     139
WMCF9-CB5:working shlee$ gzcat 2_normalforpon.vcf.gz | grep -v '#' | wc -l
     139
```

Probing further we see:
```
WMCF9-CB5:working shlee$ diff <(gzcat 1_normalforpon.vcf.gz | grep -v '#' | cut -f1,2) <(gzcat 3_discard_practice_pon.vcf.gz | grep -v '#' | cut -f1,2)
139d138
< chrX	153909841
WMCF9-CB5:working shlee$ gzcat 1_normalforpon.vcf.gz | grep 'chrX\t153909841'
chrX	153909841	.	C	A	.	.	DP=11;ECNT=1;POP_AF=1.000e-03;P_GERMLINE=-2.169e-04;TLOD=14.94	GT:AD:AF:ALT_F1R2:ALT_F2R1:FOXOG:MBQ:MCL:MFRL:MMQ:MPOS:REF_F1R2:REF_F2R1:SA_MAP_AF:SA_POST_PROB	0/1:6,5:0.455:3:2:0.400:30,33:0,0:191,278:60,60:11,20:1:5:0.404,0.444,0.455:0.025,0.025,0.950
WMCF9-CB5:working shlee$ gzcat 2_normalforpon.vcf.gz | grep 'chrX\t153909841'
chrX	153909841	.	C	A	.	.	DP=11;ECNT=1;POP_AF=1.000e-03;P_GERMLINE=-2.169e-04;TLOD=14.94	GT:AD:AF:ALT_F1R2:ALT_F2R1:FOXOG:MBQ:MCL:MFRL:MMQ:MPOS:REF_F1R2:REF_F2R1:SA_MAP_AF:SA_POST_PROB	0/1:6,5:0.455:3:2:0.400:30,33:0,0:191,278:60,60:11,20:1:5:0.404,0.444,0.455:0.025,0.025,0.950
WMCF9-CB5:working shlee$ gzcat 3_discard_practice_pon.vcf.gz | grep 'chrX'
##contig=<ID=chrX,length=156040895>
##contig=<ID=chrX_KI270880v1_alt,length=284869>
##contig=<ID=chrX_KI270881v1_alt,length=144206>
##contig=<ID=chrX_KI270913v1_alt,length=274009>
chrX	132097402	.	TACAC	T,TAC	.	.	.
```
This site should have been called in the PoN. Finally, for the `-vcfs` parameter, if I provide a list of files, one per line, the tool errors with
```
htsjdk.tribble.TribbleException$MalformedFeatureFile: Unable to parse header with error: Your input file has a malformed header: We never saw the required CHROM header line (starting with one #) for the input VCF file, for input source: /Users/shlee/Desktop/August2017_tutorial_dev/working/list_of_normals_for_pon.txt
	at htsjdk.tribble.TribbleIndexedFeatureReader.readHeader(TribbleIndexedFeatureReader.java:253)
	at htsjdk.tribble.TribbleIndexedFeatureReader.<init>(TribbleIndexedFeatureReader.java:101)
	at htsjdk.tribble.TribbleIndexedFeatureReader.<init>(TribbleIndexedFeatureReader.java:126)
	at htsjdk.tribble.AbstractFeatureReader.getFeatureReader(AbstractFeatureReader.java:110)
	at htsjdk.tribble.AbstractFeatureReader.getFeatureReader(AbstractFeatureReader.java:74)
	at htsjdk.variant.vcf.VCFFileReader.<init>(VCFFileReader.java:58)
	at org.broadinstitute.hellbender.tools.walkers.mutect.CreateSomaticPanelOfNormals.doWork(CreateSomaticPanelOfNormals.java:122)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:116)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:173)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:192)
	at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:131)
	at org.broadinstitute.hellbender.Main.mainEntry(Main.java:152)
	at org.broadinstitute.hellbender.Main.main(Main.java:233)
Caused by: htsjdk.tribble.TribbleException$InvalidHeader: Your input file has a malformed header: We never saw the required CHROM header line (starting with one #) for the input VCF file
	at htsjdk.variant.vcf.VCFCodec.readActualHeader(VCFCodec.java:115)
	at htsjdk.tribble.AsciiFeatureCodec.readHeader(AsciiFeatureCodec.java:83)
	at htsjdk.tribble.AsciiFeatureCodec.readHeader(AsciiFeatureCodec.java:36)
	at htsjdk.tribble.TribbleIndexedFeatureReader.readHeader(TribbleIndexedFeatureReader.java:251)
	... 12 more
```

Either we should update the documentation to point users to use the `--arguments_file` instead (let me know) or we should update the functionality to accept a list of files. Currently, the option is described as:

> --vcfsListFile,-vcfs:File     VCFs for samples to include. May be specified either one at a time, or as one or more .list file containing multiple VCFs, one per line.  This argument must be specified at least once. Required. 

Here is the test data:
[createsomaticpanelofnormals_testcase_shlee.zip](https://github.com/broadinstitute/gatk/files/1251993/createsomaticpanelofnormals_testcase_shlee.zip)

--- a/src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/CreateSomaticPanelOfNormals.java
+++ b/src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/CreateSomaticPanelOfNormals.java
@@ -18,6 +18,7 @@ import org.broadinstitute.barclay.help.DocumentedFeature;
 import org.broadinstitute.hellbender.cmdline.CommandLineProgram;
 import org.broadinstitute.hellbender.cmdline.StandardArgumentDefinitions;
 import org.broadinstitute.hellbender.cmdline.programgroups.VariantFilteringProgramGroup;
+import org.broadinstitute.hellbender.exceptions.UserException;
 import org.broadinstitute.hellbender.tools.walkers.haplotypecaller.AssemblyBasedCallerUtils;
 import org.broadinstitute.hellbender.utils.SimpleInterval;
 import org.broadinstitute.hellbender.utils.Utils;
@@ -52,8 +53,8 @@ import java.util.*;
  *   -O normal1_for_pon.vcf.gz
  * </pre>
  *
- * <p>Step 2. Create a file ending with .list extension with the paths to the VCFs from step 1, one per line.
- * This approach is optional. </p>
+ * <p>Step 2. Create a file ending with .args extension with the paths to the VCFs from step 1, one per line.
+ * This approach is optional.  It will fail if a file with an extension other than .args is used. </p>
  *
  * <pre>
  *     normal1_for_pon.vcf.gz
@@ -65,7 +66,7 @@ import java.util.*;
  *
  * <pre>
  * gatk CreateSomaticPanelOfNormals \
- *   -vcfs normals_for_pon_vcf.list \
+ *   -vcfs normals_for_pon_vcf.args \
  *   -O pon.vcf.gz
  * </pre>
  *
@@ -78,7 +79,10 @@ import java.util.*;
  *   -O pon.vcf.gz
  * </pre>
  *
- *  <p>The tool also accepts multiple .list files. Pass each in with the -vcfs option.</p>
+ *  <p>The tool also accepts multiple .args files. Pass each in with the -vcfs option.</p>
+ *
+ *  <p>By default the tool fails if multiple vcfs have the same sample name, but the --duplicate-sample-strategy argument can be changed to
+ *  ALLOW_ALL to allow duplicates or CHOOSE_FIRST to use only the first vcf with a given sample name.</p>
  *
  *  <p>See {@link Mutect2} documentation for usage examples.</p>
  *
@@ -92,17 +96,30 @@ import java.util.*;
 @BetaFeature
 public class CreateSomaticPanelOfNormals extends CommandLineProgram {
 
-    public static final String INPUT_VCFS_LIST_LONG_NAME = "vcfsListFile";
+    public static final String INPUT_VCFS_LIST_LONG_NAME = "vcfs";
     public static final String INPUT_VCFS_LIST_SHORT_NAME = "vcfs";
 
+    public static final String DUPLICATE_SAMPLE_STRATEGY_LONG_NAME = "duplicate-sample-strategy";
+
+    public enum DuplicateSampleStrategy {
+        THROW_ERROR, CHOOSE_FIRST, ALLOW_ALL
+    }
+
     /**
-     * The VCFs can be input as either one or more .list file(s) containing one VCF per line, or VCFs can be
+     * The VCFs can be input as either one or more .args file(s) containing one VCF per line, or VCFs can be
      * specified explicitly on the command line.
      */
     @Argument(fullName = INPUT_VCFS_LIST_LONG_NAME,
             shortName = INPUT_VCFS_LIST_SHORT_NAME,
-            doc="VCFs for samples to include. May be specified either one at a time, or as one or more .list file containing multiple VCFs, one per line.", optional = false)
+            doc="VCFs for samples to include. May be specified either one at a time, or as one or more .args file containing multiple VCFs, one per line.", optional = false)
     private Set<File> vcfs = new LinkedHashSet<>(0);
+
+    /**
+     * How to handle duplicate samples: THROW_ERROR to fail, CHOOSE_FIRST to use the first vcf with each sampel name, ALLOW_ALL to use all samples regardless of duplicate sample names."
+     */
+    @Argument(fullName = DUPLICATE_SAMPLE_STRATEGY_LONG_NAME,
+            doc="How to handle duplicate samples: THROW_ERROR to fail, CHOOSE_FIRST to use the first vcf with each sample name, ALLOW_ALL to use all samples regardless of duplicate sample names.", optional = false)
+    private DuplicateSampleStrategy duplicateSampleStrategy = DuplicateSampleStrategy.THROW_ERROR;
 
     @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
             shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
@@ -111,19 +128,32 @@ public class CreateSomaticPanelOfNormals extends CommandLineProgram {
 
     public Object doWork() {
         final List<File> inputVcfs = new ArrayList<>(vcfs);
+        if (!inputVcfs.stream().map(File::getAbsolutePath).allMatch(path -> path.endsWith(".vcf") || path.endsWith(".vcf.gz") || path.endsWith(".args") )) {
+            logger.warn("Some input files don't seem to be .vcf or .args files.  Make sure that any input vcfs list end in .args.");
+        }
         final Collection<CloseableIterator<VariantContext>> iterators = new ArrayList<>(inputVcfs.size());
         final Collection<VCFHeader> headers = new HashSet<>(inputVcfs.size());
         final VCFHeader headerOfFirstVcf = new VCFFileReader(inputVcfs.get(0), false).getFileHeader();
         final SAMSequenceDictionary sequenceDictionary = headerOfFirstVcf.getSequenceDictionary();
         final VariantContextComparator comparator = headerOfFirstVcf.getVCFRecordComparator();
 
-
+        final Set<String> samples = new HashSet<>();
         for (final File vcf : inputVcfs) {
             final VCFFileReader reader = new VCFFileReader(vcf, false);
-            iterators.add(reader.iterator());
+
             final VCFHeader header = reader.getFileHeader();
-            Utils.validateArg(comparator.isCompatible(header.getContigLines()), () -> vcf.getAbsolutePath() + " has incompatible contigs.");
-            headers.add(header);
+            final String sample = header.getGenotypeSamples().get(0);
+            if (duplicateSampleStrategy == DuplicateSampleStrategy.THROW_ERROR && samples.contains(sample)) {
+                throw new UserException.BadInput(String.format("Duplicate sample name %s found in multiple input vcfs, the second one being %s.  Consider changing the %s argument", sample, vcf.getAbsolutePath(), DUPLICATE_SAMPLE_STRATEGY_LONG_NAME));
+            } else if (duplicateSampleStrategy == DuplicateSampleStrategy.CHOOSE_FIRST && samples.contains(sample)) {
+                logger.info(String.format("Skipping input vcf %s because a different vcf with the same sample %s has already been seen", vcf.getAbsolutePath(), sample));
+                continue;
+            } else {
+                Utils.validateArg(comparator.isCompatible(header.getContigLines()), () -> vcf.getAbsolutePath() + " has incompatible contigs.");
+                headers.add(header);
+                iterators.add(reader.iterator());
+                samples.add(sample);
+            }
         }
 
         final VariantContextWriter writer = GATKVariantContextUtils.createVCFWriter(outputVcf, sequenceDictionary, false, Options.INDEX_ON_THE_FLY);
@@ -141,6 +171,7 @@ public class CreateSomaticPanelOfNormals extends CommandLineProgram {
             }
             variantsAtThisPosition.add(vc);
         }
+        processVariantsAtSamePosition(variantsAtThisPosition, writer);
         mergingIterator.close();
         writer.close();
 

--- a/src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/CreateSomaticPanelOfNormalsIntegrationTest.java
+++ b/src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/CreateSomaticPanelOfNormalsIntegrationTest.java
@@ -82,5 +82,4 @@ public class CreateSomaticPanelOfNormalsIntegrationTest extends CommandLineProgr
         Assert.assertEquals(vc5.getNAlleles(), 2);
         Assert.assertTrue(vc5.getAlternateAllele(0).basesMatch("C"));
     }
-    
 }

--- a/src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/CreateSomaticPanelOfNormalsIntegrationTest.java
+++ b/src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/CreateSomaticPanelOfNormalsIntegrationTest.java
@@ -62,7 +62,8 @@ public class CreateSomaticPanelOfNormalsIntegrationTest extends CommandLineProgr
         final File outputVcf = createTempFile("pon", ".vcf");
         final String[] args = {
                 "-" + CreateSomaticPanelOfNormals.INPUT_VCFS_LIST_SHORT_NAME, vcfInputFile.getAbsolutePath(),
-                "-O", outputVcf.getAbsolutePath()
+                "-O", outputVcf.getAbsolutePath(),
+                "--duplicate-sample-strategy", "ALLOW_ALL",
         };
 
         runCommandLine(args);
@@ -81,5 +82,5 @@ public class CreateSomaticPanelOfNormalsIntegrationTest extends CommandLineProgr
         Assert.assertEquals(vc5.getNAlleles(), 2);
         Assert.assertTrue(vc5.getAlternateAllele(0).basesMatch("C"));
     }
-
+    
 }

